kernel/kernel-api/src/main/java/io/delta/kernel/internal/snapshot/MetadataCleanup.java

```diff
@@ -0,0 +1,196 @@
+          // is potential candidate to delete later if we find another checkpoint
+          potentialLogFilesToDelete.addAll(lastSeenCheckpointFiles);
+          lastSeenCheckpointFiles.clear();
+          lastSeenCheckpointVersion = -1;
+        }
+
+        FileStatus nextFile = files.next();
+
+        // Step 2: If the timestamp is earlier than the retention period, stop
+        if (nextFile.getModificationTime() > fileCutOffTime) {
+          if (!potentialLogFilesToDelete.isEmpty()) {
+            logger.info(
+                "{}: Skipping deletion of expired log files {}, because there is no checkpoint "
+                    + "file that indicates that the log files are no longer needed. ",
+                tablePath,
+                potentialLogFilesToDelete.size());
+          }
+          break;
+        }
+
+        if (FileNames.isCommitFile(nextFile.getPath())) {
+          // Step 3: If the file is a delta log file, add it to the `potentialFilesToDelete` list
+          // We can't delete these files until we encounter a checkpoint later that indicates
+          // that the log files are no longer needed.
+          potentialLogFilesToDelete.add(nextFile.getPath());
+        } else if (FileNames.isCheckpointFile(nextFile.getPath())) {
+          // Step 4: If the file is a checkpoint file, add it to the `lastSeenCheckpointFiles`
+          long newLastSeenCheckpointVersion = FileNames.checkpointVersion(nextFile.getPath());
+          checkArgument(
+              lastSeenCheckpointVersion == -1
+                  || newLastSeenCheckpointVersion >= lastSeenCheckpointVersion);
+
+          if (lastSeenCheckpointVersion != -1
+              && newLastSeenCheckpointVersion > lastSeenCheckpointVersion) {
+            // We have found checkpoint file for a new version. This means the files gathered for
+            // the last checkpoint version are not complete (most likely an incomplete multipart
+            // checkpoint). We should delete the files gathered so far and start fresh
+            // last seen checkpoint state
+            logger.info(
+                "{}: Incomplete checkpoint files found at version {}, ignoring the checkpoint"
+                    + " files and adding them to potential log file delete list",
+                tablePath,
+                lastSeenCheckpointVersion);
+            potentialLogFilesToDelete.addAll(lastSeenCheckpointFiles);
+            lastSeenCheckpointFiles.clear();
+          }
+
+          lastSeenCheckpointFiles.add(nextFile.getPath());
+          lastSeenCheckpointVersion = newLastSeenCheckpointVersion;
+        }
+        // TODO: do we need to delete unknown file types?
+      }
+    }
+    logger.info("{}: Deleted {} log files older than {}", tablePath, numDeleted, fileCutOffTime);
+  }
+
+  private static CloseableIterator<FileStatus> listExpiredDeltaLogs(Engine engine, Path tablePath)
```

Comment:
this method doesn't do anything to ensure that the results are indeed `expired`.

`listDeltaLogs` is better -> but do we already have a nicely-named-util that does that somewhere?

Reply:
As of now, it is not filtering out the expired because it is not updating the timestamps to be monotonically increasing. I changed it to `listDeltaLogs`. There is no utility method as it depends on the prefix and the actually API call isn't that complicated.